### PR TITLE
BUG Fix delete / unpublish

### LINKED ISSUE #134: Archived and Unpublished pages do not remove 'Live' variant from index
To replicate:

Create a new page and save it as draft.
Search for the page (in incognito mode) - 0 results.

Publish the page.
Search for the page - 1 result.

Unpublish the page.
Search for the page.
Expected - 0 results.
Actual - PaginatedList reports total size of 1, but is empty.

Publish the page, then Archive it.
Search for the page.
Expected - 0 results.
Actual - PaginatedList reports total size of 1, but is empty.

Versioned alters this query so that we only get the item if it has a live version; the item is being returned from the index still, but is not being added to the list.
https://github.com/silverstripe/silverstripe-fulltextsearch/blob/master/code/solr/SolrIndex.php#L749

However, the total including the ones filtered out is being set as the PaginatedList total:
https://github.com/silverstripe/silverstripe-fulltextsearch/blob/master/code/solr/SolrIndex.php#L786

As a workaround I add these lines to affected DataObjects.

`
    public function onBeforeDelete()
    {
        $this->LastEdited = time();
        $this->write();
    }

    public function onAfterUnpublish()
    {
        singleton('nameofyourindex')->delete(
            ClassInfo::baseDataClass($this->ClassName),
            $this->ID,
            array (
              'SearchVariantVersioned' => 'Live',
              'SearchVariantSubsites' => '0',
            )
        );
    }
`

### PR DESCRIPTION
**Pull request by tractorcow**
Split SearchUpdater classes into separate files
Fixes #134

### CODE REVIEW

**Comment by tractorcow** [comment]
This works, where the existing solution didn't because:
 - Versioned nukes any manipulation against a class without a table. This means if the class only had fields in a parent class (E.g. Page, occasionally) the deletion wouldn't work. Fixed by using the entire hierarchy, so at least one table gets passed through.
 - SearchUpdater nukes any manipulation with empty fields key. Fixed by adding an special 'delete' action and having the updater exclude that from the filter.

**Comment by tractorcow** [comment]
Green. :D todo,merge into master once merge into 2.3

File: tests/SearchVariantVersionedTest.php [hunk 1]
```diff
@@ -72,6 +72,28 @@ public function testPublishing()
...
+
+        // Test unpublish
+
+        self::$index->reset();
+
+        $item->deleteFromStage('Live');
+
+        SearchUpdater::flush_dirty_indexes();
+
+        $this->assertEquals(1, count(self::$index->deleted));
```
Comment (dhensby) [inline tests/SearchVariantVersionedTest.php]: `$this->assertCount(1, self::$index->deleted)`

**Comment by dhensby** [comment]
merged into master

**Pull request merged**